### DISCUSSION TITLE
TypeError: Expected "0" to be a string

### DISCUSSION THREAD

**Discussion by J4v4Scr1pt** [vercel/next.js#72088] [Help] [answered]
### Summary

I need help to understand this error, It has started to show in our datadog logs... or flood rather.. 
Ofc this only appears in Production and not locally..

I have gone through our recent changes and reverted the ones I thought could have caused this. But no luck..

Do anyone recognize this error or maybe have an idea on why this could appear?

I know that I don't have that much to work with other than the error, but that's all I have to go on.. :/ 

Next Version: 14.2.16 (Using only app router)

![image](https://github.com/user-attachments/assets/78abdb79-6465-44dc-bebb-e2fc2d878914)

### Additional information

```js
{
at async Server.requestListener (/app/node_modules/.pnpm/next@https+++registry.npmjs.org+next+-+next-14.2.16.tgz_react-dom@18.3.1_react@18.3.1/node_modules/next/dist/server/lib/start-server.js:141:13)

at async requestHandlerImpl (/app/node_modules/.pnpm/next@https+++registry.npmjs.org+next+-+next-14.2.16.tgz_react-dom@18.3.1_react@18.3.1/node_modules/next/dist/server/lib/router-server.js:374:13)

at async handleRequest (/app/node_modules/.pnpm/next@https+++registry.npmjs.org+next+-+next-14.2.16.tgz_react-dom@18.3.1_react@18.3.1/node_modules/next/dist/server/lib/router-server.js:210:96)

at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

at resolveRoutes (/app/node_modules/.pnpm/next@https+++registry.npmjs.org+next+-+next-14.2.16.tgz_react-dom@18.3.1_react@18.3.1/node_modules/next/dist/server/lib/router-utils/resolve-routes.js:530:34)

at handleRoute (/app/node_modules/.pnpm/next@https+++registry.npmjs.org+next+-+next-14.2.16.tgz_react-dom@18.3.1_react@18.3.1/node_modules/next/dist/server/lib/router-utils/resolve-routes.js:451:94)

at prepareDestination (/app/node_modules/.pnpm/next@https+++registry.npmjs.org+next+-+next-14.2.16.tgz_react-dom@18.3.1_react@18.3.1/node_modules/next/dist/shared/lib/router/utils/prepare-destination.js:211:38)

at /app/node_modules/.pnpm/next@https+++registry.npmjs.org+next+-+next-14.2.16.tgz_react-dom@18.3.1_react@18.3.1/node_modules/next/dist/compiled/path-to-regexp/index.js:237:19

TypeError: Expected "0" to be a string
}
```

**Comment by icyJoseph** [comment]
So, I think this is coming from this library, https://github.com/pillarjs/path-to-regexp/blob/d63f44bc54f2c6fe7d35dc77f8515622006bf7cb/src/index.ts#L422, which is used to match paths and such in Next.js, mostly in middleware IIRC, https://nextjs.org/docs/app/building-your-application/routing/middleware#matcher

Could it be that someone/something is forging path that causes your middleware matcher to throw and error?

There are some tests in that library, https://github.com/pillarjs/path-to-regexp/blob/d63f44bc54f2c6fe7d35dc77f8515622006bf7cb/src/index.spec.ts#L77, which show when this error would be logged. It is like the matcher is getting an object `{"0": /* some value */}` - very weird indeed!

**Reply by J4v4Scr1pt** [reply]
Very good investigation sir! And very interesting stuff.

We do use middleware, but nothing has changed here in a long time, last change is 6 months old.
We did an update to nextjs from **`14.2.15 --> 14.2.16`** 2 days ago, but this error just showed up today.

We are also looking to implementing LaunchDarkly into Node using this [template project](https://github.com/launchdarkly-labs/ld-nextjs). They are using the instrumentation file for this. I did try to leave the instrumentation file empty by commeting out the code to see if it was this that caused the error, but nothing changed.

Our middleware looks like this:

```
export default async function middleware(request: NextRequest) {
	const hasAuthCookie= request.cookies.has('authCookie');
	if (!hasAuthCookie) return NextResponse.redirect(new URL(`${process.env.Login}`));

	const { device } = userAgent(request);
	const isMobile = device.type === 'mobile' || device.type === 'tablet';

	if (!isMobile && request.url.includes('MobileLanding'))
		return NextResponse.redirect(new URL('Home', request.url));

	if (isMobile && !request.url.includes('MobileLanding'))
		return NextResponse.redirect(new URL('MobileLanding', request.url));
}

export const config = {
	matcher: [
		{
			source: '/((?!Login|api|_next/static|_next/image|favicon.ico).*)',
			missing: [
				// Exclude Server Actions
				{ type: 'header', key: 'next-action' },
			],
		},
		'/',
	],
};
```

**Reply by J4v4Scr1pt** [reply]
I was thinking, we are migrating a application to nextjs and have them living side by side. The sidemenu we have link to both nextjs and the old application. The users can move between them seamlessly, but we do get an error in the client that the Link component can't prefetch the URL beacuse of obvious reasons that it link to an non nextjs page. Maybe this could be something.

I will re-write this logic to not use the Link component when it's not nextjs pages and see if this goes away.

**Reply by icyJoseph** [reply]
Indeed that could be a possible root cause.

Do your logs show in what route the error is triggered?

**Reply by J4v4Scr1pt** [reply]
Unfortunately the logs dont tell anything regarding which route or url that triggerd this. 
But! (Knock on wood) my fix with only render menuItems with Link component if it is nextjs page seems to have fixed this. 

I'm very grateful that you helped me investigate this! 

I will monitor the logs during the day and mark your post as answer if there is no more logs :)

**Reply by icyJoseph** [reply]
Great! Would have been nice to have a reproduction demo though, but quick production stability is more important. Good job and good luck!

**Reply by J4v4Scr1pt** [reply]
Agree, but I don't know how to create this demo though. Locally this never accure in the server logs, only in production.

The error did apear again, but I saw that we use the Link component again on our mobile landing page linking to apple/android store for our app. Replaced this with regluar a-tag. And built again, hopefully this time no logs will appear.

Because I could count the number of logs and tie it to the number of menu items in the sidebar that did link to pages not being nextjs.
The new logs where only two and that makes sense because there where two links on the mobile landing page.

Hopefully now no more logs will show 😄

**Reply by J4v4Scr1pt** [reply]
I think we can put this to done. I have tested everything over and over like I did before.
No new logs for over 2h.

Thank you again for helping me out with this!

Link component is a no go with URLs other than Nextjs pages. 👍

**Reply by icyJoseph** [reply]
Nice. I wonder if the Middleware is getting hit for those external links, somehow? Well, good that you caught it, hopefully if others get it too, they'll be able to quickly fix it.

**Reply by J4v4Scr1pt** [reply]
I apparently forgot to knock on wood.. the error reaped again. 
Need to investigate more..

**Reply by icyJoseph** [reply]
It has to be a malformed route hitting your Middleware :/ 🤔

**Reply by J4v4Scr1pt** [reply]
I will spend my day tomorrow testing until this appear to understand why.

**Reply by icyJoseph** [reply]
Do you have dynamic routes in your project?

**Reply by J4v4Scr1pt** [reply]
> Do you have dynamic routes in your project?

All routes are dynamic atm. 

I have been testing all morning doing different stuff and still no error. Gone through the code and there is no Link component usage anywhere except NextJs routes.

**Reply by icyJoseph** [reply]
I saw that the prepareDestination file is on the error stack.. there they talk about routes that look like, `/foo/:param` 🤔

**Reply by icyJoseph** [reply]
See here, https://github.com/vercel/next.js/blob/canary/packages%2Fnext%2Fsrc%2Fshared%2Flib%2Frouter%2Futils%2Fprepare-destination.ts#L200

So I wonder if a bot, or some other actor, is firing requests against a malformed URL... As said before... Too bad the logs don't show which route ~ might be worth dropping a temporary log on the Middleware, saying something like... Processing this url... And then try to correlate that to the issue we see

**Reply by J4v4Scr1pt** [reply]
Very interesting indeed!
I will drop a log as you suggested and keep this until the error appears again and hopefully there is some valuable information.

`console.log("🚀 ~ Processing  ~ URL:", request.url)`

**Reply by icyJoseph** [reply]
Maybe put it behind an env variable or something, depends... If it's easy to redeploy a removal... Because that one will pollute your logs heavily, but it is for science 🧪🔭

**Reply by J4v4Scr1pt** [reply]
> Maybe put it behind an env variable or something, depends... If it's easy to redeploy a removal... Because that one will pollute your logs heavily, but it is for science 🧪🔭

Haha, yes for science 🙌 😅!  
Thanks for the heads up 🙏, it's quick and easy to re-deploy and remove it :)

**Reply by mikib0** [reply]
I'm using next 14 (app router) and a config I'd forgotten about was the culprit:
```
exportPathMap: async function (
    defaultPathMap,
    { dev, dir, outDir, distDir, buildId }
  ) {
    return {
      '/': { page: '/(app)/(onboarding)/pitch' },
    };
  },
```
Removing it fixed it for me. Wishing you luck😆

**Reply by J4v4Scr1pt** [reply]
> I'm using next 14 (app router) and a config I'd forgotten about was the culprit:
> 
> ```
> exportPathMap: async function (
>     defaultPathMap,
>     { dev, dir, outDir, distDir, buildId }
>   ) {
>     return {
>       '/': { page: '/(app)/(onboarding)/pitch' },
>     };
>   },
> ```
> 
> Removing it fixed it for me. Wishing you luck😆

Thx for this suggestion! I will go through and check if there could be anything like this 🙏

**Reply by J4v4Scr1pt** [reply]
Ok I figured out a consistent way of triggering the error. 
To have our old application living side by side with our new we use the nextConfig to have redirects to pages that have not been migrated to next yet from the sideMenu. They live under the same domain/url except that we utilize the basePath setting and have all NextJs pages under `/next`. 

The logic I implemented as I talked about above was to render the menuItem as a regular a-tag if it's not a nextjs page otherwise render the Link component.
When using the sideMenu the redirect works as expected, you can go between pages without any issues and you will be navigated back and forth. Super great!

Worth mentioning, we have a `not-found.tsx` file to handle any unknown routes and a `error.tsx` on every route including `global-error.tsx`. This has worked as expected.

However here is the interesting part, if we have a button to push you to the old application with the nextjs hook `useRouter` like this `onPress={() => router.push('/URL-WITH-REDIRECT')}` the page crashes totally and you end up on the Image below. Bypassing the `global-error.tsx` I reckon.. This also happens if I enter the URL in the Browsers url-field directly. But here's the kicker if I put an url that is not in the redirect Array in the nextConfig we end up on the not-found page. This goes for both router.push and manual enter url.

Next highly interesting thing is that locally it works 🤯!! In other words this is only happening in the Production build.

**The middleware do not! log when this happens!** It logs if I enter a non-existing url that ends up on the not-found page. 

![image](https://github.com/user-attachments/assets/0929f7bf-da07-4e61-97b9-85f0cff06a4c)

**Reply by icyJoseph** [reply]
Does this happen locally when you do `next build && next start` - need that clarification.

Also, are you able to make a small reproduction repository, with only the minimal stuff needed to cause the issue?

**Reply by J4v4Scr1pt** [reply]
I will build this locally and test!
I can create an Stackblitz?

But I think I found the solution, now reading the docs again:
[https://nextjs.org/docs/app/api-reference/next-config-js/redirects#redirects-with-basepath-support](https://nextjs.org/docs/app/api-reference/next-config-js/redirects#redirects-with-basepath-support)

Will test this first.

**Reply by J4v4Scr1pt** [reply]
I'm stupid, ofc it works when I use a regular a-tag pointing to an url without the base path... 
The router.push did what it should, but I guess there was some loop happening crashing the site. Because a redirect without the basePath setting will redirect back to /next and then round it goes.

**Reply by icyJoseph** [reply]
This was an interesting debugging session! So, do you think you are close to landing a fix? Would be great to see if a stackblitz yeah, just to document the steps that lead to such a cryptic error message

**Reply by J4v4Scr1pt** [reply]
I found another issue! ohh man.. Sometimes you wonder if you are fit for doing programming.. 😝..

I don't think a stackblitz is needed.. 

We.. I.. had created the redirects looking like this:
	          
                       {
				source: '/Url',
				destination: `${process.env.OLD_APP_BASE_PATH}Url`,
				permanent: false,
			},
			
Is it possible to access environment variables in the nextConfig.. ~~no ofc not~~..(yes, see below) then it ended up looking like this after the build:

                   {
				source: '/Url',
				destination: `Url`,
				permanent: false,
			},

Which will cause the issue and explains all the errors, because the Link component tried to pre-fetch on the url that would redirect wrong etc.

Now I don't know how to solve the issue with the redirect instead... 😢 
I have tried with and without the basePath: false, setting on the redirect config. But I end up with the basePath anyways. 

Because we want to redirect them to the same url but without the basePath.
e.g `https://app.companyName.com/next/some-page --> https://app.companyName.com/some-page `
Most of the pages will keep the url and to have them living on the same domain and url the basePath was a good solution.

**Reply by icyJoseph** [reply]
Try using the function form of the next config object export:

- https://nextjs.org/docs/app/api-reference/next-config-js#configuration-as-a-function

But, as far as I know you should definitely have access to the env vars, are you making them available at build time?

**Reply by J4v4Scr1pt** [reply]
> Try using the function form of the next config object export:
> 
> * https://nextjs.org/docs/app/api-reference/next-config-js#configuration-as-a-function
> 
> But, as far as I know you should definitely have access to the env vars, are you making them available at build time?

ok.. hmm I googled and there was some answer on StackOverflow saying I would not.. But ok then I will test with this again. We are going off topic with this and will test around with it and open another thread if I get stuck.

Like I said before in the thread @icyJoseph you have been an rock in this investigation and I thank you very very much for that 🌟!
I will write a summary post below this and mark that as answer.

Edit: Just tried logging and you are correct I can access the env variables in the config. Thanks!
Edit2: The environment variable was wrong during the build that's why it became empty. Now that is fixed and everything works 🙌

**Reply by J4v4Scr1pt** [reply, marked as answer]
The reason for the error was incorrect/faulty usage of redirects in the nextjs config causing circular behavior with the usage of the Link component. This also affected the usage of router.push and manually entering the redirected URL in the browser URL-field. Make sure your redirects are built and look correct before usage.
